### ISSUE TITLE
Need the ability to supply a custom synchronization context

### ISSUE THREAD

**Issue by simonholt** [zeromq/netmq#699]
While considering porting an application from NetMQ 3 to NetMQ 4, I came across a scenario (in our unit test framework) where I need to supply a non-standard SynchronizationContext;

In NetMQ3, this was quite simple:

                using (var syncContext = new UnitTestSynchronizationContext(scheduler))
                {
                    SynchronizationContext.SetSynchronizationContext(syncContext);
                    socketPoller.PollTillCancelled();
                }

In NetMQ4, we can't do this because NetMQPoller.Run always creates and uses its own SynchronizationContext.  I'd like to create an overload of NetMQPoller.Run that accepts a SynchronizationContext for this purpose.

I imagine this is a pretty niche case so I don't expect there to be much demand for it; with that in mind, are you happy for me to submit a pull request for this?

**Comment by simonholt** [comment]
I've created pull request #700 with the sort of thing I have in mind.  Please let me know what you think.

**Comment by somdoron** [comment]
Pull request is welcomed... I saw you already close it. Anyway feel free to submit it.

**Comment by simonholt** [comment]
I haven't discovered the cause of my test problem yet, but I have ruled out the change I proposed as the cause.

It seems that calling:

   NetMQ.NetMQConfig.Cleanup(true);

causes a hang in 4.0.0.4-beta, but works fine in 4.0.0-rc5.  It seems to wait in Ctx at the following line forever:

   var found = m_termMailbox.TryRecv(-1, out Command command);

I'll spend some time trying to narrow down more precisely where and what the cause of this is over the next couple of days.
